src/gui/transferlistmodel.h

```diff
@@ -102,6 +109,9 @@ private slots:
 private:
     QList<BitTorrent::TorrentHandle *> m_torrentList;  // maps row number to torrent handle
     QHash<BitTorrent::TorrentHandle *, int> m_torrentMap;  // maps torrent handle to row number
+
+    // row text colors
+    std::unordered_map<BitTorrent::TorrentState, QColor> m_stateForegroundColors;
```

Comment:
Let's try using QHash here, what was the error message? FYI I can use `QHash<BitTorrent::TorrentState, QColor> m_stateForegroundColors;` without providing a custom qHash() function.

Reply:
```
[9/17] Building CXX object src\gui\CMakeFiles\qbt_gui.dir\transferlistmodel.cpp.obj
FAILED: src/gui/CMakeFiles/qbt_gui.dir/transferlistmodel.cpp.obj
E:\Cpp\VS\2019\VC\Tools\MSVC\14.24.28314\bin\Hostx64\x64\cl.exe  /nologo /TP -DBOOST_ALL_NO_LIB -DNOMINMAX -DNTDDI_VERSION=0x06010000 -DQBT_VERSION=\"v4.3.0alpha1\" -DQBT_VERSION_2=\"4.3.0alpha1\" -DQBT_VERSION_BUGFIX=0 -DQBT_VERSION_BUILD=0 -DQBT_VERSION_MAJOR=4 -DQBT_VERSION_MINOR=3 -DQT_CORE_LIB -DQT_DEPRECATED_WARNINGS -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_NO_CAST_TO_ASCII -DQT_STRICT_ITERATORS -DQT_USE_QSTRINGBUILDER -DQT_WIDGETS_LIB -DQT_WINEXTRAS_LIB -DQT_XML_LIB -DSTACKTRACE -DUNICODE -DWIN32 -DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_DEPRECATE -D_UNICODE -D_WIN32 -D_WIN32_IE=0x0601 -D_WIN32_WINNT=0x0601 -Isrc\gui\qbt_gui_autogen\include -I..\src -I..\src\gui\..\app -I..\src\gui -I..\src\gui\powermanagement -I..\src\gui\properties -IE:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include -IE:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore -IE:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\.\mkspecs\win32-msvc -IE:\Cpp\Sources\libtorrent\include -IE:\Cpp\Sources\OpenSSL-Win64\include -IE:\Cpp\Sources\boost\boost_1_71_0 -IE:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtNetwork -IE:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtXml -IE:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtGui -IE:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtANGLE -IE:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtWidgets -IE:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtWinExtras /DWIN32 /D_WINDOWS /W3 /GR /EHsc -wd4290 -wd4275 -wd4251 /W4 /MDd /Zi /Ob0 /Od /RTC1   -DBOOST_ASIO_DISABLE_CONNECTEX -DBOOST_EXCEPTION_DISABLE -DTORRENT_USE_LIBCRYPTO -DTORRENT_USE_OPENSSL -DTORRENT_DISABLE_RESOLVE_COUNTRIES -DTORRENT_DEBUG -DTORRENT_LINKING_SHARED -DBOOST_SYSTEM_DYN_LINK -std:c++14 /showIncludes /Fosrc\gui\CMakeFiles\qbt_gui.dir\transferlistmodel.cpp.obj /Fdsrc\gui\CMakeFiles\qbt_gui.dir\qbt_gui.pdb /FS -c ..\src\gui\transferlistmodel.cpp
cl : Command line warning D9025 : overriding '/W3' with '/W4'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(117): error C2665: 'BitTorrent::qHash': none of the 32 overloads could convert all the argument types
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtGui/qtransform.h(205): note: could be 'uint qHash(const QTransform &,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtGui/qmatrix.h(143): note: or       'uint qHash(const QMatrix &,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore\qdatetime.h(414): note: or       'uint qHash(const QTime &,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore\qdatetime.h(413): note: or       'uint qHash(const QDate &,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore\qdatetime.h(412): note: or       'uint qHash(const QDateTime &,uint)'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qlocale.h(62): note: or       'uint qHash(const QLocale &,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore\qabstractitemmodel.h(467): note: or       'uint qHash(const QModelIndex &) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore\qabstractitemmodel.h(153): note: or       'uint qHash(const QPersistentModelIndex &,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qregexp.h(56): note: or       'uint qHash(const QRegExp &,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(107): note: or       'uint qHash(std::nullptr_t,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(104): note: or       'uint qHash(QLatin1String,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(103): note: or       'uint qHash(const QBitArray &,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(102): note: or       'uint qHash(QStringView,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(100): note: or       'uint qHash(const QStringRef &,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(99): note: or       'uint qHash(const QString &,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(97): note: or       'uint qHash(const QByteArray &,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(96): note: or       'uint qHash(const QChar,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(94): note: or       'uint qHash(long double,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(92): note: or       'uint qHash(double,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(91): note: or       'uint qHash(float,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(90): note: or       'uint qHash(qint64,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(86): note: or       'uint qHash(quint64,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(85): note: or       'uint qHash(long,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(79): note: or       'uint qHash(ulong,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(78): note: or       'uint qHash(int,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(77): note: or       'uint qHash(uint,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(76): note: or       'uint qHash(short,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(75): note: or       'uint qHash(ushort,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(74): note: or       'uint qHash(signed char,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(73): note: or       'uint qHash(uchar,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(72): note: or       'uint qHash(char,uint) noexcept'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(117): note: while trying to match the argument list '(const T)'
        with
        [
            T=BitTorrent::TorrentState
        ]
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(116): note: see reference to function template instantiation 'uint qHash<Key>(const T &,uint) noexcept(<expr>)' being compiled
        with
        [
            Key=BitTorrent::TorrentState,
            T=BitTorrent::TorrentState
        ]
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore\qhash.h(930): note: while compiling class template member function 'QHashNode<Key,T> **QHash<Key,T>::findNode(const Key &,uint *) const'
        with
        [
            Key=BitTorrent::TorrentState,
            T=QColor
        ]
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore\qhash.h(758): note: see reference to function template instantiation 'QHashNode<Key,T> **QHash<Key,T>::findNode(const Key &,uint *) const' being compiled
        with
        [
            Key=BitTorrent::TorrentState,
            T=QColor
        ]
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore\qhash.h(881): note: while compiling class template member function 'void QHash<BitTorrent::TorrentState,QColor>::reserve(int)'
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore\qhash.h(248): note: see reference to function template instantiation 'void QHash<BitTorrent::TorrentState,QColor>::reserve(int)' being compiled
E:\Cpp\Projects\bittorent\qBittorrent\src\gui\transferlistmodel.h(115): note: see reference to class template instantiation 'QHash<BitTorrent::TorrentState,QColor>' being compiled
E:\Cpp\Qt\5.13.0\5.13.0\msvc2017_64\include\QtCore/qhashfunctions.h(117): error C2056: illegal expression
ninja: build stopped: subcommand failed.
```

Reply:
Weird...
I only did 3 things in transferlistmodel.h to get it compile on gcc:
1. Add `QHash<BitTorrent::TorrentState, QColor> m_stateForegroundColors;`
2. Add `enum class TorrentState;` forward declaration.
3. Add `#include <QColor>`

Reply:
nope not working. also tried directly including base/bittorrent/torrenthandle.h

Reply:
Well, then at least add a comment about it: 
`// TODO: use QHash when MSVC doesn't complain about it anymore`

Reply:
Then why not just declare appropriate qHash function? It's trivial.
Or use underlying integer type for QHash keys?

Reply:
done

Reply:
Does using qt containers really that important. I saw some talk of one of Qt developer and he himself advised to not use Qt Containers.

Reply:
>Does using qt containers really that important.

I think No, unless there are some benefits in some particular case. Don't forget that Qt containers are Implicitly shared, for e.g..

Reply:
I considered using underlying type but then I have to use static_cast everywhere, especially using that in constructor will only produce useless noise.

Reply:
It's no problem for me to use unordered_map here.

Reply:
>Does using qt containers really that important.

I think the point is the coherency of container/data structure used through out the project. Now there is a comment stating the reason of using std container, then I'm fine with it.

>I saw some talk of one of Qt developer and he himself advised to not use Qt Containers.

I considered about it awhile ago but I think it is too much work for little/none gain for this project.

Reply:
Since std/Qt containers are not complete substitutes for each other, such a question should never be raised. We only can choose default containers for project (Qt containers currently) and use others when it has meaning.